src/inc/conpty-universal.h

```diff
@@ -30,18 +30,161 @@
+//      The sort is case-insensitive, Unicode order, without regard to locale.
+//      Because the equal sign is a separator, it must not be used in the name of
+//      an environment variable."
+//      https://docs.microsoft.com/en-us/windows/desktop/ProcThread/changing-environment-variables
+//
+struct WStringCaseInsensitiveCompare
+{
+    bool operator()(const std::wstring& lhs, const std::wstring& rhs) const
+    {
+        return (::_wcsicmp(lhs.c_str(), rhs.c_str()) < 0);
+    }
+};
+
+using EnvironmentVariableMapW = std::map<std::wstring, std::wstring, WStringCaseInsensitiveCompare>;
+
+HRESULT CreateConPty(const std::wstring& cmdline,                       // _In_
+                     const unsigned short w,                            // _In_
+                     const unsigned short h,                            // _In_
+                     HANDLE* const hInput,                              // _Out_
+                     HANDLE* const hOutput,                             // _Out_
+                     HANDLE* const hSignal,                             // _Out_
+                     PROCESS_INFORMATION* const piPty,                  // _Out_
+                     const EnvironmentVariableMapW& extraEnvVars = {}); // _In_
 
 bool SignalResizeWindow(const HANDLE hSignal,
                         const unsigned short w,
                         const unsigned short h);
 
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map);
+
+HRESULT EnvironmentMapToEnvironmentStringsW(EnvironmentVariableMapW& map, std::vector<wchar_t>& newEnvVars);
+
+// Function Description:
+// - Updates an EnvironmentVariableMapW with the current process's unicode
+//   environment variables ignoring ones already set in the provided map.
+// Arguments:
+// - map: The map to populate with the current processes's environment variables.
+// Return Value:
+// - S_OK if we succeeded, or an appropriate HRESULT for failing
+__declspec(noinline) inline
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map)
+{
+    LPWCH currentEnvVars{};
+    auto freeCurrentEnv = wil::scope_exit([&] {
+        if (currentEnvVars)
+        {
+            (void)FreeEnvironmentStringsW(currentEnvVars);
+            currentEnvVars = nullptr;
+        }
+    });
+
+    RETURN_IF_NULL_ALLOC(currentEnvVars = ::GetEnvironmentStringsW());
+
+    // Block is guaranteed to be double-NULL terminated at a minimum.
+    for (wchar_t const* lastCh{ currentEnvVars }; *lastCh != '\0';)
+    {
+        // Copy current entry into temporary map.
+        const size_t cchEntry{ ::wcslen(lastCh) };
+        const std::wstring_view entry{ lastCh, cchEntry };
+
+        // Every entry is of the form "name=value\0".
+        auto pos = entry.find_first_of(L"=", 0, 1);
+        RETURN_HR_IF(E_UNEXPECTED, pos == std::wstring::npos);
+
+        std::wstring name{ lastCh, pos };
+        std::wstring value{ lastCh + pos + 1, cchEntry - pos - 1 };

```

Comment:
all of this +1 and -1 stuff makes me nervous, is it possible to use `stringstream` here?

Reply:
stringstream is best-suited to pre-known sized data, but I have some ideas.  I'll try a few out later tonight.

Reply:
Yeah, istringstream ended up not being a great fit here as it really likes to work with pre-sized strings.  I didn't want to have to scan through the entire environment block just to figure out total length first.  I've reworked this to only use find, substr and one or two additions with no negatives instead.  Hopefully that's good enough.